File: app/static/js/game.js
class ChineseChessGame {
    constructor() {
        this.gameState = null;
        this.selectedPiece = null;
        this.validMoves = [];
        this.isAIThinking = false;
        this.init();
    }

    init() {
        this.setupEventListeners();
        this.renderGridLines();
        this.loadGameState();
        this.setupResizeListener();
    }

    getCellSize() {
        const root = document.documentElement;
        const cellSizeValue = getComputedStyle(root).getPropertyValue('--cell-size');
        return parseFloat(cellSizeValue) || 50;
    }

    renderGridLines() {
        const gridLines = document.getElementById('gridLines');
        if (!gridLines) return;

        gridLines.innerHTML = '';
        const cellSize = this.getCellSize();

        for (let row = 0; row < 10; row++) {
            const line = document.createElement('div');
            line.className = 'grid-line-h';
            line.style.top = (row * cellSize) + 'px';
            gridLines.appendChild(line);
        }

        for (let col = 0; col < 9; col++) {
            const x = col * cellSize;

            if (col === 0 || col === 8) {
                const line = document.createElement('div');
                line.className = 'grid-line-v';
                line.style.left = x + 'px';
                gridLines.appendChild(line);
            } else {
                const topLine = document.createElement('div');
                topLine.className = 'grid-line-v-top';
                topLine.style.left = x + 'px';
                topLine.style.top = '0px';
                gridLines.appendChild(topLine);

                const bottomLine = document.createElement('div');
                bottomLine.className = 'grid-line-v-bottom';
                bottomLine.style.left = x + 'px';
                bottomLine.style.top = (5 * cellSize) + 'px';
                gridLines.appendChild(bottomLine);
            }
        }

        const topPalace = { row: 0, col: 3 };
        const bottomPalace = { row: 7, col: 3 };

        [topPalace, bottomPalace].forEach(palace => {
            const y = palace.row * cellSize;
            const x = palace.col * cellSize;

            const diag1 = document.createElement('div');
            diag1.className = 'palace-diag palace-diag-1';
            diag1.style.left = x + 'px';
            diag1.style.top = y + 'px';
            gridLines.appendChild(diag1);

            const diag2 = document.createElement('div');
            diag2.className = 'palace-diag palace-diag-2';
            diag2.style.left = (x + 2 * cellSize) + 'px';
            diag2.style.top = y + 'px';
            gridLines.appendChild(diag2);
        });
    }

    setupEventListeners() {
        const newGameBtn = document.getElementById('newGameBtn');
        if (newGameBtn) {
            newGameBtn.addEventListener('click', () => this.showNewGameModal());
        }

        const shareBtn = document.getElementById('shareBtn');
        if (shareBtn) {
            shareBtn.addEventListener('click', () => this.shareGame());
        }

        const refreshBtn = document.getElementById('refreshBtn');
        if (refreshBtn) {
            refreshBtn.addEventListener('click', () => this.loadGameState());
        }

        const modalClose = document.querySelectorAll('.modal-close');
        modalClose.forEach(btn => {
            btn.addEventListener('click', (e) => {
                e.target.closest('.modal').classList.remove('show');
            });
        });

        document.querySelectorAll('.modal').forEach(modal => {
            modal.addEventListener('click', (e) => {
                if (e.target === modal) {
                    modal.classList.remove('show');
                }
            });
        });

        this.setupColorButtons();
        this.setupDifficultyButtons();

        const startBtn = document.getElementById('startGameBtn');
        if (startBtn) {
            startBtn.addEventListener('click', () => this.startNewGame());
        }

        const copyBtn = document.getElementById('copyShareBtn');
        if (copyBtn) {
            copyBtn.addEventListener('click', () => this.copyShareLink());
        }
    }

    setupColorButtons() {
        const colorBtns = document.querySelectorAll('.color-btn');
        colorBtns.forEach(btn => {
            btn.addEventListener('click', () => {
                colorBtns.forEach(b => b.classList.remove('selected'));
                btn.classList.add('selected');
            });
        });
    }

    setupDifficultyButtons() {
        const diffBtns = document.querySelectorAll('.difficulty-btn');
        diffBtns.forEach(btn => {
            btn.addEventListener('click', () => {
                diffBtns.forEach(b => b.classList.remove('selected'));
                btn.classList.add('selected');
            });
        });
    }

    setupResizeListener() {
        let resizeTimeout;
        window.addEventListener('resize', () => {
            clearTimeout(resizeTimeout);
            resizeTimeout = setTimeout(() => {
                this.renderGridLines();
                if (this.gameState) {
                    this.renderBoard();
                }
            }, 100);
        });
    }

    showNewGameModal() {
        const modal = document.getElementById('newGameModal');
        if (modal) {
            modal.classList.add('show');
        }
    }

    async startNewGame() {
        const selectedColor = document.querySelector('.color-btn.selected');
        const selectedDiff = document.querySelector('.difficulty-btn.selected');

        const playerColor = selectedColor ? selectedColor.dataset.color : 'red';
        const difficulty = selectedDiff ? selectedDiff.dataset.difficulty : 'medium';

        try {
            const response = await fetch('/api/start', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ player_color: playerColor, difficulty })
            });

            const data = await response.json();
            this.gameState = data;
            this.renderBoard();
            this.updateGameInfo();

            const modal = document.getElementById('newGameModal');
            if (modal) modal.classList.remove('show');
        } catch (error) {
            console.error('Failed to start game:', error);
        }
    }

    async loadGameState() {
        try {
            const response = await fetch('/api/state');
            const data = await response.json();

            if (data.status !== 'waiting') {
                this.gameState = data;
                this.renderBoard();
                this.updateGameInfo();
            } else {
                this.showNewGameModal();
            }
        } catch (error) {
            console.error('Failed to load game state:', error);
            this.showNewGameModal();
        }
    }

    renderBoard() {
        const piecesLayer = document.querySelector('.pieces-layer');
        if (!piecesLayer || !this.gameState) return;

        piecesLayer.innerHTML = '';

        const board = this.gameState.board;
        const cellSize = this.getCellSize();

        for (let row = 0; row < 10; row++) {
            for (let col = 0; col < 9; col++) {
                const piece = board[row][col];
                if (piece) {
                    const x = col * cellSize;
                    const y = row * cellSize;

                    const cell = document.createElement('div');
                    cell.className = 'intersection';
                    cell.style.left = x + 'px';
                    cell.style.top = y + 'px';
                    cell.dataset.row = row;
                    cell.dataset.col = col;

                    const pieceEl = document.createElement('div');
                    pieceEl.className = `piece ${piece.color}`;
                    pieceEl.textContent = piece.name;

                    if (this.gameState.selected_piece &&
                        this.gameState.selected_piece[0] === row &&
                        this.gameState.selected_piece[1] === col) {
                        pieceEl.classList.add('selected');
                    }

                    cell.appendChild(pieceEl);
                    cell.addEventListener('click', () => this.handleCellClick(row, col));
                    piecesLayer.appendChild(cell);
                }
            }
        }

        this.renderValidMoves();
        this.renderLastMove();
    }

    renderValidMoves() {
        const piecesLayer = document.querySelector('.pieces-layer');
        if (!piecesLayer || !this.gameState) return;

        const validMoves = this.gameState.valid_moves || [];
        const cellSize = this.getCellSize();

        validMoves.forEach(([row, col]) => {
            const x = col * cellSize;
            const y = row * cellSize;

            const targetPiece = this.gameState.board[row][col];

            const clickArea = document.createElement('div');
            clickArea.className = 'clickable-area';
            clickArea.style.left = x + 'px';
            clickArea.style.top = y + 'px';
            clickArea.addEventListener('click', () => this.handleCellClick(row, col));
            piecesLayer.appendChild(clickArea);

            const moveEl = document.createElement('div');
            if (targetPiece) {
                moveEl.className = 'valid-capture';
            } else {
                moveEl.className = 'valid-move';
            }
            moveEl.style.left = x + 'px';
            moveEl.style.top = y + 'px';
            piecesLayer.appendChild(moveEl);
        });
    }

    renderLastMove() {
        const piecesLayer = document.querySelector('.pieces-layer');
        if (!piecesLayer || !this.gameState) return;

        const lastMove = this.gameState.last_move;
        if (!lastMove) return;

        const cellSize = this.getCellSize();
        const [fromRow, fromCol, toRow, toCol] = lastMove;

        [[fromRow, fromCol], [toRow, toCol]].forEach(([row, col]) => {
            const x = col * cellSize;
            const y = row * cellSize;

            const moveEl = document.createElement('div');
            moveEl.className = 'last-move-highlight';
            moveEl.style.left = x + 'px';
            moveEl.style.top = y + 'px';
            piecesLayer.appendChild(moveEl);
        });
    }

    async handleCellClick(row, col) {
        if (this.isAIThinking) return;
        if (!this.gameState) return;

        const status = this.gameState.status;
        if (status !== 'playing' && status !== 'check') {
            return;
        }

        if (this.gameState.current_player !== this.gameState.player_color) {
            return;
        }

        const clickedPiece = this.gameState.board[row][col];
        const validMoves = this.gameState.valid_moves || [];
        const isInValidMoves = validMoves.some(m => m[0] === row && m[1] === col);

        if (clickedPiece && clickedPiece.color === this.gameState.player_color) {
            await this.selectPiece(row, col);
        } else if (isInValidMoves && this.gameState.selected_piece) {
            const [fromRow, fromCol] = this.gameState.selected_piece;
            await this.makeMove(fromRow, fromCol, row, col);
        }
    }

    async selectPiece(row, col) {
        try {
            const response = await fetch('/api/select', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ row, col })
            });

            const data = await response.json();
            if (data.success) {
                this.gameState = data.state;
                this.renderBoard();
                this.updateGameInfo();
            }
        } catch (error) {
            console.error('Failed to select piece:', error);
        }
    }

    async makeMove(fromRow, fromCol, toRow, toCol) {
        try {
            const response = await fetch('/api/move', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ from_row: fromRow, from_col: fromCol, to_row: toRow, to_col: toCol })
            });

            const data = await response.json();
            if (data.success) {
                this.gameState = data.state;
                this.renderBoard();
                this.updateGameInfo();

                this.checkGameEnd();
            }
        } catch (error) {
            console.error('Failed to make move:', error);
        }
    }

    updateGameInfo() {
        if (!this.gameState) return;

        const turnIndicator = document.querySelector('.turn-dot');
        if (turnIndicator) {
            turnIndicator.className = `turn-dot ${this.gameState.current_player}`;
        }

        const turnText = document.getElementById('turnText');
        if (turnText) {
            turnText.textContent = this.gameState.current_player === 'red' ? '红方回合' : '黑方回合';
        }

        const statusText = document.getElementById('statusText');
        if (statusText) {
            const statusMap = {
                'waiting': '等待开始',
                'playing': '游戏进行中',
                'check': '将军！',
                'checkmate': '将死！',
                'stalemate': '困毙！',
                'finished': '游戏结束'
            };
            statusText.textContent = statusMap[this.gameState.status] || this.gameState.status;
            statusText.className = 'status-text';
            if (this.gameState.status === 'check') {
                statusText.classList.add('check');
            }
        }

        const difficultyText = document.getElementById('difficultyText');
        if (difficultyText) {
            const diffMap = { 'easy': '简单', 'medium': '中等', 'hard': '困难' };
            difficultyText.textContent = `难度: ${diffMap[this.gameState.difficulty] || this.gameState.difficulty}`;
        }

        this.updateMoveHistory();
    }

    updateMoveHistory() {
        const moveList = document.getElementById('moveList');
        if (!moveList || !this.gameState) return;

        moveList.innerHTML = '';
        const notations = this.gameState.move_notations || [];

        notations.forEach((notation, index) => {
            const li = document.createElement('li');
            li.className = 'move-item';

            const moveNum = document.createElement('span');
            moveNum.className = 'move-number';
            moveNum.textContent = `${index + 1}.`;

            const moveNotation = document.createElement('span');
            moveNotation.className = 'move-notation';
            moveNotation.textContent = notation;

            li.appendChild(moveNum);
            li.appendChild(moveNotation);
            moveList.appendChild(li);
        });

        moveList.scrollTop = moveList.scrollHeight;
    }

    checkGameEnd() {
        if (!this.gameState) return;

        if (this.gameState.status === 'checkmate') {
            setTimeout(() => {
                const winner = this.gameState.winner === 'red' ? '红方' : '黑方';
                alert(`游戏结束！${winner}获胜！`);
            }, 100);
        } else if (this.gameState.status === 'stalemate') {
            setTimeout(() => {
                alert('游戏结束！困毙！');
            }, 100);
        }
    }

    async shareGame() {
        if (!this.gameState || this.gameState.status === 'waiting') {
            alert('请先开始游戏！');
            return;
        }

        try {
            const response = await fetch('/api/share');
            const data = await response.json();

            const shareUrl = document.getElementById('shareUrl');
            const shareCode = document.getElementById('shareCode');

            if (shareUrl) shareUrl.value = data.share_url;
            if (shareCode) shareCode.value = data.share_code;

            const modal = document.getElementById('shareModal');
            if (modal) modal.classList.add('show');
        } catch (error) {
            console.error('Failed to share game:', error);
            alert('分享失败，请重试');
        }
    }

    async copyShareLink() {
        const shareUrl = document.getElementById('shareUrl');
        if (shareUrl) {
            try {
                await navigator.clipboard.writeText(shareUrl.value);
                alert('分享链接已复制到剪贴板！');
            } catch (error) {
                shareUrl.select();
                document.execCommand('copy');
                alert('分享链接已复制到剪贴板！');
            }
        }
    }
}

document.addEventListener('DOMContentLoaded', () => {
    new ChineseChessGame();
});

File: app/chess/game.py
from typing import Optional, List, Tuple
from enum import Enum
from app.chess.board import Board
from app.chess.pieces import Color, Piece
from app.chess.ai import AI, Difficulty
from app.chess.notation import Notation


class GameStatus(Enum):
    WAITING = 'waiting'
    PLAYING = 'playing'
    CHECK = 'check'
    CHECKMATE = 'checkmate'
    STALEMATE = 'stalemate'
    FINISHED = 'finished'


class Game:
    def __init__(self):
        self.board = Board()
        self.current_player: Color = Color.RED
        self.status: GameStatus = GameStatus.WAITING
        self.player_color: Color = Color.RED
        self.ai_color: Color = Color.BLACK
        self.difficulty: Difficulty = Difficulty.MEDIUM
        self.ai: Optional[AI] = None
        self.move_notations: List[str] = []
        self.selected_piece: Optional[Tuple[int, int]] = None
        self.valid_moves: List[Tuple[int, int]] = []
        self.last_move: Optional[Tuple[int, int, int, int]] = None
        self.winner: Optional[Color] = None

    def start_game(self, player_color: Color = Color.RED, difficulty: Difficulty = Difficulty.MEDIUM):
        self.board = Board()
        self.current_player = Color.RED
        self.status = GameStatus.PLAYING
        self.player_color = player_color
        self.ai_color = Color.BLACK if player_color == Color.RED else Color.RED
        self.difficulty = difficulty
        self.ai = AI(self.ai_color, difficulty)
        self.move_notations = []
        self.selected_piece = None
        self.valid_moves = []
        self.last_move = None
        self.winner = None

        if self.ai_color == Color.RED:
            self._make_ai_move()

    def select_piece(self, row: int, col: int) -> bool:
        if self.status not in [GameStatus.PLAYING, GameStatus.CHECK]:
            return False

        if self.current_player != self.player_color:
            return False

        piece = self.board.get_piece(row, col)
        if piece and piece.color == self.current_player:
            self.selected_piece = (row, col)
            self.valid_moves = list(self.board.get_valid_moves(row, col))
            return True

        return False

    def make_move(self, from_row: int, from_col: int, to_row: int, to_col: int) -> bool:
        if self.status not in [GameStatus.PLAYING, GameStatus.CHECK]:
            return False

        if self.current_player != self.player_color:
            return False

        valid_moves = self.board.get_valid_moves(from_row, from_col)
        if (to_row, to_col) not in valid_moves:
            return False

        notation = Notation.move_to_notation(self.board, from_row, from_col, to_row, to_col)

        self.board.move_piece(from_row, from_col, to_row, to_col)
        self.move_notations.append(notation)
        self.last_move = (from_row, from_col, to_row, to_col)
        self.selected_piece = None
        self.valid_moves = []

        if not self._check_game_end():
            self.current_player = self.ai_color
            self._make_ai_move()

        return True

    def _make_ai_move(self):
        if self.ai is None:
            return

        if self.status not in [GameStatus.PLAYING, GameStatus.CHECK]:
            return

        move = self.ai.get_best_move(self.board)
        if move is None:
            return

        from_row, from_col, to_row, to_col = move
        notation = Notation.move_to_notation(self.board, from_row, from_col, to_row, to_col)

        self.board.move_piece(from_row, from_col, to_row, to_col)
        self.move_notations.append(notation)
        self.last_move = (from_row, from_col, to_row, to_col)

        self._check_game_end()

        if self.status in [GameStatus.PLAYING, GameStatus.CHECK]:
            self.current_player = self.player_color

    def _check_game_end(self) -> bool:
        opponent_color = Color.BLACK if self.current_player == Color.RED else Color.RED

        if self.board.is_in_check(opponent_color):
            if self.board.is_checkmate(opponent_color):
                self.status = GameStatus.CHECKMATE
                self.winner = self.current_player
                return True
            else:
                self.status = GameStatus.CHECK
        elif self.board.is_stalemate(opponent_color):
            self.status = GameStatus.STALEMATE
            return True
        else:
            self.status = GameStatus.PLAYING

        return False

    def get_game_state(self) -> dict:
        board_state = []
        for row in range(Board.ROWS):
            row_data = []
            for col in range(Board.COLS):
                piece = self.board.get_piece(row, col)
                if piece:
                    row_data.append({
                        'row': row,
                        'col': col,
                        'color': piece.color.value,
                        'type': piece.piece_type.value,
                        'name': piece.name
                    })
                else:
                    row_data.append(None)
            board_state.append(row_data)

        return {
            'board': board_state,
            'current_player': self.current_player.value,
            'player_color': self.player_color.value,
            'status': self.status.value,
            'difficulty': self.difficulty.value,
            'move_notations': self.move_notations,
            'selected_piece': self.selected_piece,
            'valid_moves': self.valid_moves,
            'last_move': self.last_move,
            'winner': self.winner.value if self.winner else None
        }

    def get_share_code(self) -> str:
        initial_board = Board()
        notations = Notation.encode_game_history(self.board.move_history, initial_board)

        config = f"{self.player_color.value}|{self.difficulty.value}"
        return f"{config}|{notations}"

    @staticmethod
    def from_share_code(share_code: str) -> Optional['Game']:
        try:
            parts = share_code.split('|', 2)
            if len(parts) < 3:
                return None

            player_color_str, difficulty_str, notations_str = parts

            player_color = Color(player_color_str)
            difficulty = Difficulty(difficulty_str)

            game = Game()
            game.player_color = player_color
            game.ai_color = Color.BLACK if player_color == Color.RED else Color.RED
            game.difficulty = difficulty
            game.ai = AI(game.ai_color, difficulty)

            notations = Notation.decode_game_history(notations_str)
            game.move_notations = notations.copy()

            temp_board = Board()
            current_color = Color.RED

            for notation in notations:
                moves = Notation.notation_to_moves(temp_board, notation, current_color)
                if moves:
                    from_row, from_col, to_row, to_col = moves[0]
                    temp_board.move_piece(from_row, from_col, to_row, to_col)
                    game.last_move = (from_row, from_col, to_row, to_col)
                    current_color = Color.BLACK if current_color == Color.RED else Color.RED

            game.board = temp_board
            game.current_player = current_color

            if game.board.is_in_check(game.current_player):
                if game.board.is_checkmate(game.current_player):
                    game.status = GameStatus.CHECKMATE
                    game.winner = Color.BLACK if game.current_player == Color.RED else Color.RED
                else:
                    game.status = GameStatus.CHECK
            elif game.board.is_stalemate(game.current_player):
                game.status = GameStatus.STALEMATE
            else:
                game.status = GameStatus.PLAYING

            return game
        except Exception:
            return None

    def load_replay(self, share_code: str) -> List[Tuple[str, dict]]:
        try:
            parts = share_code.split('|', 2)
            if len(parts) < 3:
                return []

            player_color_str, difficulty_str, notations_str = parts
            notations = Notation.decode_game_history(notations_str)

            replay_steps = []
            temp_board = Board()
            current_color = Color.RED

            initial_state = self._get_board_state(temp_board)
            replay_steps.append(('初始局面', initial_state))

            for i, notation in enumerate(notations):
                moves = Notation.notation_to_moves(temp_board, notation, current_color)
                if moves:
                    from_row, from_col, to_row, to_col = moves[0]
                    temp_board.move_piece(from_row, from_col, to_row, to_col)

                    step_name = f"第{i+1}步: {notation}"
                    board_state = self._get_board_state(temp_board)
                    board_state['last_move'] = (from_row, from_col, to_row, to_col)
                    replay_steps.append((step_name, board_state))

                    current_color = Color.BLACK if current_color == Color.RED else Color.RED

            return replay_steps
        except Exception:
            return []

    def _get_board_state(self, board: Board) -> dict:
        board_state = []
        for row in range(Board.ROWS):
            row_data = []
            for col in range(Board.COLS):
                piece = board.get_piece(row, col)
                if piece:
                    row_data.append({
                        'row': row,
                        'col': col,
                        'color': piece.color.value,
                        'type': piece.piece_type.value,
                        'name': piece.name
                    })
                else:
                    row_data.append(None)
            board_state.append(row_data)
        return {'board': board_state, 'last_move': None}
